mlir/lib/Dialect/Rock/Transforms/LowerRockReduce.cpp

```diff
@@ -0,0 +1,233 @@
+  redInput =
+      rewriter.create<TransformOp>(loc, redInput, threadsToInpTensorAttr);
+
+  threadsToInpTensor =
+      BottomUpTMBuilder::above(threadsToInpTensor, threadsToInpTensorAttr);
+  threadsToInpTensor.pad({"flatDim"},
+                         {0, totalThreads * dataPerThread - topSizeProduct});
+  threadsToInpTensorAttr = threadsToInpTensor.get();
+  redInput =
+      rewriter.create<TransformOp>(loc, redInput, threadsToInpTensorAttr);
+
+  threadsToInpTensor =
+      BottomUpTMBuilder::above(threadsToInpTensor, threadsToInpTensorAttr);
+  threadsToInpTensor.unmerge({"bid", "tid", "iter"}, {0, 1, 2}, "flatDim",
+                             {gridSize, blockSize, dataPerThread});
+  TransformMapAttr tidIterAttr = threadsToInpTensor.get();
+  Value transformed = rewriter.create<TransformOp>(loc, redInput, tidIterAttr);
+
+  return transformed;
+}
+
+static LogicalResult getStoreMethod(ReduceMethod rMethod,
+                                    StoreMethod &stMethod) {
+  if (rMethod == ReduceMethod::Sum) {
+    stMethod = StoreMethod::AtomicAdd;
+    return success();
+  }
+  return failure();
+}
+
+LogicalResult ReduceRewritePattern::matchAndRewrite(
+    ReduceOp op, ReduceOpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+  auto loc = op->getLoc();
+  uint64_t redAxis = adaptor.getAxisAttr().getInt();
+  auto gridSize = adaptor.getGridSizeAttr().getInt();
+  auto blockSize = adaptor.getBlockSizeAttr().getInt();
+
+  // This pass doing this rewrite means the preceding
+  // fusion passes did not perform any fusion and the reduction
+  // op will become the anchor op. Therefore, setting the
+  // blockSize and gridSize attr of the func.
+  auto funcOp = op->getParentOfType<func::FuncOp>();
+  if (funcOp->hasAttr("block_size")) {
+    llvm::errs() << "blockSize = " << funcOp->getAttr("block_size") << "\n";
+    return op.emitError("func op should not have a blockSize set for a "
+                        "standalone rock.reduce op");
+  }
+  if (funcOp->hasAttr("grid_size")) {
+    return op.emitError("func op should not have a gridSize set for a "
+                        "standalone rock.reduce op");
+  }
+  funcOp->setAttr("block_size", adaptor.getBlockSizeAttr());
+  funcOp->setAttr("grid_size", adaptor.getGridSizeAttr());
+
+  auto threadViewVal =
+      createThreadView(adaptor.getIn(), blockSize, gridSize, loc, rewriter);
+  ArrayAttr sourceTransformsFromOp;
+  Value source;
+  std::tie(source, sourceTransformsFromOp) =
+      untransform(rewriter, threadViewVal);
+  auto loadOobDims = computeOobFromTransforms(rewriter, sourceTransformsFromOp);
+
+  auto threadViewShape = threadViewVal.getType().cast<ShapedType>().getShape();
+  SmallVector<int64_t> bounds(threadViewShape.size(), 1LL);
+  bounds[threadViewShape.size() - 1] =
+      threadViewShape[threadViewShape.size() - 1];
+  SmallVector<int64_t> strides(threadViewShape.size(), 1LL);
+
+  // Get current workgroup ID.
```

Comment:
High-level suggestion, instead of using `(bid, tid, i) -> flatten()` use `(bid, i, tid)` so that we get well-behaved coalescing loads.

Reply:
Done!